pycon_archive_past_website/pycon_archive_past_website.py

```diff
@@ -269,43 +42,32 @@ def get_page(path):
             link.get("href").find("talk") != -1
             or link.get("href").find("tutorial") != -1
         ):
-            talk_link = unquote(link["href"])
-            get_page(talk_link)
+            get_page(crawler, unquote(link["href"]))
 
 
-def main():
-    # Get pycon website, including zh-hant and en-us, according to given year.
-    request = requests.get(PYCON_URL + "/" + PYCON_YEAR + "/zh-hant/")  # Get HTML
-    soup = BeautifulSoup(request.text, "html.parser")  # Using html parser
-    crawler_urls = soup.select("a")
-    crawler_urls = set([crawler_url["href"] for crawler_url in crawler_urls])
-    if PYCON_YEAR >= "2020":
-        request = requests.get(
-            PYCON_URL + "/" + PYCON_YEAR + "/zh-hant/events/warmup-session/"
-        )
-        soup = BeautifulSoup(request.text, "html.parser")
-        crawler_urls_2020_warmup_session = soup.select("a")
-        for url in crawler_urls_2020_warmup_session:
-            url = url["href"]
-            crawler_urls.add(url)
-        crawler_urls.add(f"/{PYCON_YEAR}/zh-hant/sponsor/prospectus/")
+def main(year: str, base_url: str):
+    crawler: BaseCrawler = CRAWLERS[year]("https://tw.pycon.org", base_url)
+    crawler_urls = crawler.get_crawl_urls()
 
+    # Page crawler section
     for crawler_url in crawler_urls:
         url = urlparse(crawler_url)
         # Checking if the url is a pycon website
-        if url.netloc != PYCON_HOST and url.netloc != "":
+        if url.netloc != crawler.host and url.netloc != "":
             continue
         # Checking if the path is right or not
-        if url.netloc == "" and url.path.find(f"/{PYCON_YEAR}") != 0:
+        if url.netloc == "" and url.path.find(f"/{crawler.year}") != 0:
             continue
         path_parts = Path(url.path).parts
-        if len(path_parts) >= 2 and path_parts[1] == PYCON_YEAR:
-            get_page(url.path)
-            get_page(url.path.replace("zh-hant", "en-us"))
+        if len(path_parts) >= 2 and path_parts[1] == crawler.year:
+            get_page(crawler, crawler_url)
+            get_page(crawler, crawler_url.replace("zh-hant", "en-us"))
 
+    # Get favicon on front page
+    soup = get_soup(f"{crawler.url}/{crawler.year}/zh-hant/")
     for link in soup.findAll("link", {"rel": "icon"}):
         if "href" in link.attrs:
-            get_assets(Path(urlparse(link["href"]).path))
+            get_asset(link["href"])
```

Comment:
I guess this part is also the execution belonging to the crawler that filters out the urls and decides which one should be scraped. I guess moving it into the `BaseCrawler` will look cleaner.

Reply:
Moved L66~L70 to separate `BaseCrawler` class and rename it to `get_favicon` now.